lib/Demangling/Remangler.cpp

```diff
@@ -441,9 +441,11 @@ std::pair<int, Node *> Remangler::mangleConstrainedType(Node *node,
     Chain.push_back(node->getChild(1), Factory);
     node = getChildOfType(node->getFirstChild());
   }
-  
+
   if (node->getKind() != Node::Kind::DependentGenericParamType) {
     mangle(node, depth + 1);
+    if (!Chain.size())
+      return {-1, nullptr};
```

Comment:
This bit slightly concerns me. It makes sense, and it passes all the tests, but I'm not 100% confident of it. I *think* it's right because if there are no dependent member types and no dependent generic parameter types, it can't be a constrained type and should be rejected.

Without the `if (!Chain.size())`, we break `Remangler::mangleDependentMemberType`, where it's clearly legal for the returned `Node *` to be `nullptr`. And without the `return` statement here at all, we can be tripped up in `Remangler::mangleDependentGenericConformanceRequirement()`, `Remangler::mangleDependentGenericSameTypeRequirement()` or `Remangler::mangleDependentGenericLayoutRequirement()`; the test vectors that accompany these changes trigger the assertion failure in the second of those.

Reply:
There are no dependent member types that don't eventually terminate at a generic parameter, so this `if` makes sense for detecting a mangle tree that describes a type that is structurally impossible in a well-formed program.

Reply:
Thanks for confirming that. 🙂